.github/workflows/ci.yml

```diff
@@ -15,32 +15,55 @@ jobs:
   compile:
     name: Compile
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        target:
+          - riscv32imc-esp-espidf
+          - xtensa-esp32-espidf
+          - xtensa-esp32s2-espidf
+          - xtensa-esp32s3-espidf
     steps:
       - name: Setup | Checkout
         uses: actions/checkout@v2
+
       - name: Setup | Rust
         uses: actions-rs/toolchain@v1
         with:
           toolchain: ${{ env.rust_toolchain }}
           components: rustfmt, clippy
+          default: true
+        if: matrix.target == 'riscv32imc-esp-espidf'
+
       - name: Setup | Std
         run: rustup component add rust-src --toolchain ${{ env.rust_toolchain }}-x86_64-unknown-linux-gnu
-      - name: Setup | Default to nightly
-        run: rustup default ${{ env.rust_toolchain }}
+        if: matrix.target == 'riscv32imc-esp-espidf'
+
+      - name: Install Rust for Xtensa
+        uses: esp-rs/xtensa-toolchain@v1.2.0
+        with:
+          default: true
+        if: matrix.target != 'riscv32imc-esp-espidf'
+
       - name: Build | Fmt Check
         run: cargo fmt -- --check
+
       - name: Build | Clippy
-        run: export ESP_IDF_SDKCONFIG_DEFAULTS=$(pwd)/.github/configs/sdkconfig.defaults; cargo clippy --no-deps --target riscv32imc-esp-espidf -Zbuild-std=std,panic_abort -Zbuild-std-features=panic_immediate_abort -- -Dwarnings 
+        run: export ESP_IDF_VERSION=release/v4.4; export ESP_IDF_SDKCONFIG_DEFAULTS=$(pwd)/.github/configs/sdkconfig.defaults; cargo clippy --no-deps --target ${{ matrix.target }} -Zbuild-std=std,panic_abort -Zbuild-std-features=panic_immediate_abort -- -Dwarnings
+
       - name: Build | Compile
-        run: export ESP_IDF_SDKCONFIG_DEFAULTS=$(pwd)/.github/configs/sdkconfig.defaults; cargo build --target riscv32imc-esp-espidf -Zbuild-std=std,panic_abort -Zbuild-std-features=panic_immediate_abort
+        run: export ESP_IDF_VERSION=release/v4.4; export ESP_IDF_SDKCONFIG_DEFAULTS=$(pwd)/.github/configs/sdkconfig.defaults; cargo build --target ${{ matrix.target }} -Zbuild-std=std,panic_abort -Zbuild-std-features=panic_immediate_abort
+
       - name: Build | RISCV-ULP-HAL feature
-        run: cargo build --features riscv-ulp-hal --no-default-features --target riscv32imc-unknown-none-elf -Zbuild-std=core,panic_abort -Zbuild-std-features=panic_immediate_abort
+        run: cargo build --features riscv-ulp-hal --no-default-features --target ${{ matrix.target }} -Zbuild-std=core,panic_abort -Zbuild-std-features=panic_immediate_abort
```

Comment:
I am a bit confused. What target is this supposed to have? `xtensa-esp32s2-espidf`, `riscv32imc-esp-espidf`, `${{ matrix.target }}` or something else?

Also judging from lib.rs:
```
#[cfg(all(feature = "riscv-ulp-hal", not(esp32s2)))]
compile_error!("Feature `ulp` is currently only supported on esp32s2");
```
we are supposed to be targeting the esp32s2 device, but compiling for its riscv-coprocessor? What flags should we use to tell cargo this?

----

Also, perhaps semi unrelated, looking at [IDF-doc](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/hw-reference/chip-series-comparison.html) both S2 and S3 have this thing? Would it be as trivial as simply enable it for the S3 aswell?

Reply:
I think we can just remove this part I think. The RISC-V ulp hal has to be compiled with `riscv32imc-unknown-none-elf` so no point adding this test to the xtensa target matrix.